plugins/GeoIp2/GeoIP2AutoUpdater.php

```diff
@@ -0,0 +1,647 @@
+     *                    from this URL.
+     * @throws Exception
+     */
+    protected function downloadFile($dbType, $url)
+    {
+        $url = trim($url);
+
+        $ext = GeoIP2AutoUpdater::getGeoIPUrlExtension($url);
+
+        // NOTE: using the first item in $dbNames[$dbType] makes sure GeoLiteCity will be renamed to GeoIPCity
+        $zippedFilename = LocationProviderGeoIp2::$dbNames[$dbType][0] . '.' . $ext;
+
+        $zippedOutputPath = LocationProviderGeoIp2::getPathForGeoIpDatabase($zippedFilename);
+
+        $url = self::removeDateFromUrl($url);
+
+        // download zipped file to misc dir
+        try {
+            $success = Http::sendHttpRequest($url, $timeout = 3600, $userAgent = null, $zippedOutputPath);
+        } catch (Exception $ex) {
+            throw new Exception("GeoIP2AutoUpdater: failed to download '$url' to "
+                . "'$zippedOutputPath': " . $ex->getMessage());
+        }
+
+        if ($success !== true) {
+            throw new Exception("GeoIP2AutoUpdater: failed to download '$url' to "
+                . "'$zippedOutputPath'! (Unknown error)");
+        }
+
+        Log::info("GeoIP2AutoUpdater: successfully downloaded '%s'", $url);
+
+        try {
+            self::unzipDownloadedFile($zippedOutputPath, $dbType, $unlink = true);
+        } catch (Exception $ex) {
+            throw new Exception("GeoIP2AutoUpdater: failed to unzip '$zippedOutputPath' after "
+                . "downloading " . "'$url': " . $ex->getMessage());
+        }
+
+        Log::info("GeoIP2AutoUpdater: successfully updated GeoIP 2 database '%s'", $url);
+    }
+
+    /**
+     * Unzips a downloaded GeoIP 2 database. Only unzips .gz & .tar.gz files.
+     *
+     * @param string $path Path to zipped file.
+     * @param bool $unlink Whether to unlink archive or not.
+     * @throws Exception
+     */
+    public static function unzipDownloadedFile($path, $dbType, $unlink = false)
+    {
+        $tempPath = LocationProviderGeoIp2::getPathForGeoIpDatabase('update/');
+
+        // extract file
+        if (substr($path, -7, 7) == '.tar.gz') {
+            $unzip = Unzip::factory('tar.gz', $path);
+            $success = $unzip->extract($tempPath);
+
+            if ($success !== true) {
+                throw new Exception(Piwik::translate('UserCountry_CannotUnzipDatFile',
+                    array("'$path'", $unzip->errorInfo())));
+            }
+        } else if (substr($path, -3, 3) == '.gz') {
+            $unzip = Unzip::factory('gz', $path);
+            $success = $unzip->extract($tempPath);
+
+            if ($success !== true) {
+                throw new Exception(Piwik::translate('UserCountry_CannotUnzipDatFile',
+                    array("'$path'", $unzip->errorInfo())));
+            }
+        } else {
+            $ext = end(explode(basename($path), '.', 2));
+            throw new Exception(Piwik::translate('UserCountry_UnsupportedArchiveType', "'$ext'"));
+        }
```

Comment:
It would be possible for tar.gz files. All other unzip classes do not have a possibility to list and extract specific content. My problem here is, that I do not know if archives of paid databases do contain the same structure and have the same archive type. The current implementation is most flexible, as it extracts everything and then checks if any extracted file matches a database name...

Reply:
agreed here, @sgiehl can we delete all files that don't match a white-listed extension? this would help make sure no php or other executable files could be downloaded by an attacker on the server.

Reply:
From what I can tell, the updater only supports .tar.gz & .gz files, and .gz archives can only have a single file, so filtering them out for the .tar.gz file would be enough.

Reply:
I'll check that later

Reply:
I've now changed the extraction method to only extract the first file with a valid geoip database name